telegram/_utils/warnings_transition.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+#
+# A library that provides a Python interface to the Telegram Bot API
+# Copyright (C) 2015-2023
+# Leandro Toledo de Souza <devs@python-telegram-bot.org>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains classes used for transition warnings issued by this library.
+
+It was created to prevent circular imports that would be caused by creating the warnings
+inside warnings.py.
+
+.. versionadded:: NEXT.VERSION
+"""
+import functools
+from collections.abc import Callable
+from typing import Any, Tuple
+
+from telegram._utils.warnings import warn
+from telegram.warnings import PTBDeprecationWarning
+
+
+# Narrower type hints will cause linting errors and/or circular imports.
+# We'll use `Any` here and put type hints in the calling code.
```

Comment:
Well, I don't know... In this cases I can't say:
```py
self.thumbnail_height: Optional[int]
```
in class because mypy will complain about incompatible types. And I wouldn't want `self.thumbnail_height` to be of type `Union[PhotoSize, str, int, None]`.

Maybe I'm misunderstanding something

Reply:
What I meant was something like 

```python
def warn_about_deprecated_arg_return_new_arg(
    deprecated_arg: _T,
    new_arg: _S,
    deprecated_arg_name: str,
    new_arg_name: str,
    bot_api_version: str,
    stacklevel: int = 3,
) -> Union[_T, _S]:
```

but apparently this doesn't play well with the decision to stick with `arg: type = None` instead of converting to `arg: Optional[type] = None` (can't find where we decidide that right now). Anyways, I don't think it's important enough to follow this up any further :)